### PR TITLE
Add Factory Save Feature

### PR DESCRIPTION
**Pull request by tvwenger**
Add a feature to download the factory plan as JSON (or some other format).

### CODE REVIEW

**Comment by tvwenger** [comment]
Implements saving of factory to JSON, loading a factory from JSON, and adding new production lines onto an existing factory.

File: src/factory.ts [hunk 1]
```diff
@@ -410,8 +410,11 @@ function sanityCheck(factory: FactoryGraph): void {
  */
 export function buildFactory(
     requirements: Map<Craftable, { count: number; maintain: number }>,
+    factory: FactoryGraph | undefined,
```
Comment (lgfrbcsgo) [inline src/factory.ts:413]: Optional parameters can be declared using `?`
```suggestion
    factory?: FactoryGraph,
```

File: src/graph.ts [hunk 2]
```diff
@@ -628,9 +631,12 @@ export class FactoryGraph {
      * Add a container to the factory graph
      * @see {@link ContainerNode}
      */
-    createContainer(item: Item): ContainerNode {
+    createContainer(item: Item, id: string | undefined = undefined): ContainerNode {
```
Comment (lgfrbcsgo) [inline src/graph.ts:634]:
```suggestion
    createContainer(item: Item, id?: string): ContainerNode {
```

File: src/graph.ts [hunk 3]
```diff
@@ -639,9 +645,16 @@ export class FactoryGraph {
      * Add a split container to the factory graph
      * @see {@link ContainerNode}
      */
-    createSplitContainer(item: Item, split: number): ContainerNode {
+    createSplitContainer(
+        item: Item,
+        split: number,
+        id: string | undefined = undefined,
```
Comment (lgfrbcsgo) [inline src/graph.ts:651]:
```suggestion
        id?: string,
```

File: src/graph.ts [hunk 4]
```diff
@@ -650,12 +663,14 @@ export class FactoryGraph {
...
      */
-    createTransferContainer(items: Item[]): TransferContainerNode {
-        const container = new TransferContainerNode(
-            `Trans Container${this.transferContainers.size}`,
-            items,
-            this,
-        )
+    createTransferContainer(
+        items: Item[],
+        id: string | undefined = undefined,
```
Comment (lgfrbcsgo) [inline src/graph.ts:668]:
```suggestion
        id?: string,
```

File: src/graph.ts [hunk 5]
```diff
@@ -664,15 +679,17 @@ export class FactoryGraph {
      * Add an output node to the factory graph
      * @see {@link OutputNode}
      */
-    createOutput(item: Craftable, outputRate: PerMinute, maintainedOutput: Quantity): OutputNode {
+    createOutput(
+        item: Craftable,
+        outputRate: PerMinute,
+        maintainedOutput: Quantity,
+        id: string | undefined = undefined,
```
Comment (lgfrbcsgo) [inline src/graph.ts:686]:
```suggestion
        id?: string,
```

File: src/items.ts [hunk 6]
```diff
@@ -218,7 +218,9 @@ export function containerElement(
 /**
  * Item list (TODO: complete)
  */
-export const ITEMS = {
+export const ITEMS: {
+    [key: string]: any
+} = {
```
Comment (lgfrbcsgo) [inline src/items.ts:223]: Why is this change necessary? The inferred type is stronger then the declared type here.

Comment (tvwenger) [reply src/items.ts]: You are correct! Thanks for the fix. Implemented in https://github.com/lgfrbcsgo/du-factory-generator/commit/2660050714c6a7c71b8c8f3d3fa6d292a40db9eb

**Comment by tvwenger** [comment]
I made the suggested changes. I added a new feature:

When starting from an existing factory and adding more assemblers for an item that's already in the factory's output, send the output of the new assemblers to the same output node if possible.

I also fixed two bugs:

1. Adding production lines for catalysts can add upstream production of byproducts. Since byproduct handling happened before catalyst handling, there were some byproducts that weren't being handled. This is fixed now.

2. There was a bug with how container egress was being calculated for TransferNode consumers outputing to a TransferContainerNode. Before, I was assuming that the TransferNode drew equally from each input. This is a poor assumption because some input containers might be fed by more industries than another input container. Now, I keep track of what the container's egress to the industry was before it was moved into the TransferNode and TransferContainerNode.

File: src/graph.ts [hunk 7]
```diff
@@ -511,38 +515,43 @@ export class TransferNode {
...
             for (const consumer of this.output.consumers) {
-                if (consumer instanceof IndustryNode) {
+                if (isIndustryNode(consumer)) {
+                    // add the industry consumer input rate
                     consumptionRate += consumer.getInput(this.item)
                 } else {
-                    consumptionRate += consumer.inflowRatePerContainer
+                    // the only time we get here is when the output's consumer is a transfer unit,
+                    // in which case it is simply moving byproduct.
+                    consumptionRate += 0
```
Comment (lgfrbcsgo) [inline src/graph.ts:530]: Is this intentional? That's a no-op.

**Review by lgfrbcsgo — APPROVED** [review]

File: src/graph.ts (hunk 7 above)
Comment (tvwenger) [reply src/graph.ts:530]: Yes, just to make it clear that the movement of byproduct is not counted in the egress of a container. We may wish to change this in the future.
